doc/making-a-visualizer.md

```diff
@@ -1,17 +1,194 @@
+one of the basic examples as you read.
+
+#### :key: Name _(required)_
+
+Put the visualizer's name in the static `visualizationName` property. The name
+is used in the visualizer list and the titles of bundle cards.
+
+#### :bulb: Parameters _(often used)_
+
+Parameters are the user-facing structures that ask for values when a
+visualizer is created. When the user clicks "save", each parameter stores its
+value in a corresponding top-level property of the visualizer object. If you
+change one of those top-level properties while the visualizer is running, you
+should refresh the corresponding parameter by calling `this.refreshParams()`.
+
+Below the list of parameter properties, the visualizer class has a `params`
+property that describes how the parameters should appear in the UI. Look in
+`src/shared/Paramable.ts`, or in other visualizers, to learn about the options
+you can set in the `params` property.
+
+-   **p5 Visualizer Template:** `stepSize = 1`.
+-   **Differences:** `n = 20`, `levels = 5`. The parameters are refreshed in
+    the in the `inhabit()` method.
+
+#### :bulb: Other top-level properties _(often used)_
+
+You may also need top-level properties that are set and updated while the
+visualizer is running, beyond the user's direct control. By convention, we
+list these properties after the `params` property.
+
+-   **p5 Visualizer Template:** `index`, `flash`.
+-   **Differences:** `first`.
+
+#### :bulb: Check parameters _(often used)_
+
+When the user clicks "save", `checkParameters()` is called, giving you a
+chance to check the parameter values and prompt the user to correct any
+invalid ones.
+
+Your validation can't depend on the sequence, because there might not be any
+sequence loaded when the validation check runs. You can do sequence-dependent
+validation in [`inhabit()`](#advanced-inhabit), as described below, but you
+won't be able to prompt the user for corrections at that point.
+
+-   **p5 Visualizer Template:** Make sure that the step size is positive.
+-   **Differences:** Make sure that the number of terms is at least the number
+    of levels.
+
+#### :nut*and_bolt: Inhabit a page element *(advanced)\_
+
+Each time the visualizer is inserted into a page element, the `inhabit()`
+function is called, giving you access to the element the visualizer is about
+to inhabit. If you implement `inhabit()`, start by passing the given element
+up to `super.inhabit()`, which does the behind-the-scenes work of managing p5
+canvass.
+
+When `inhabit()` is called, the visualizer will always be attached to a
+sequence, so you can do sequence-dependent validation and initialization here.
+
+-   **p5 Visualizer Template:** Go to the beginning of the sequence.
+-   **Differences:** Fill in the default value of the `levels` property in
+    case it was left unset in the parameters dialog. Do a consistency check
+    between the sequence being visualized and the parameters.
+
+#### :bulb: Set up the visualizer _(often used)_
+
```

Comment:
If you go with the idea to move `inhabit()` out of this section, the bulk of the comments can be shifted down to here, as this is where we're now recommending doing all of that stuff.

Reply:
Good point! How's the organization now?

Reply:
The guide for what setup() is and does seems fine, but the inhabit() section is still there and still seems to have outdated information. If you are convinced there should be an inhabit() section, fine, but make sure everything in it is correct. And you may want to add info along the lines of "Unless you need information from the surrounding web page or need to manipulate its document object model (DOM), you probably don't need an `inhabit()` method."

Reply:
Whoops, now I see the outdated info in the `inhabit()` section now. I've removed it.

> And you may want to add info along the lines of "Unless you need information from the surrounding web page or need to manipulate its document object model (DOM), you probably don't need an inhabit() method."

Done.

Reply:
Yup, that's fine. Only thing left is the remaining inhabit() section is now one of the longest in this tutorial part, which is not good for a method we doubt "beginners" will even want or need to implement. Suggest moving all but the first two-sentence paragraph to the more technical abstract visualizers section, or eliminating altogether if they don't seem relevant there.

Reply:
Yeah, that looks better. Done.

Reply:
Yup, all seems reasonable now. Resolving.